variables.tf

```diff
@@ -1,41 +1,133 @@
+    If this is false and no egress rules are specified via `rules` or `rule-matrix`, then no egress will be allowed.
+    EOT
+}
+
 variable "rules" {
   type        = list(any)
-  default     = null
+  default     = []
   description = <<-EOT
-    A list of maps of Security Group rules. 
-    The values of map is fully complated with `aws_security_group_rule` resource. 
+    A list of maps of Security Group rules.
+    The keys and values of the maps are fully compatible with the `aws_security_group_rule` resource, except
+    for `security_group_id` which will be ignored.
     To get more info see https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule .
   EOT
 }
+
+variable "rule_matrix" {
+  # rule_matrix is independent of the `rules` input.
+  # Only the rules specified in the `rule_matrix` object are applied to the subjects specified in `rule_matrix`.
+  #  Schema:
+  #  {
+  #    # these top level lists define all the subjects to which rule_matrix rules will be applied
+  #    source_security_group_ids = list of source security group IDs to apply all rules to
+  #    cidr_blocks = list of ipv4 CIDR blocks to apply all rules to
+  #    ipv6_cidr_blocks = list of ipv6 CIDR blocks to apply all rules to
+  #    prefix_list_ids = list of prefix list IDs to apply all rules to
+  #    self = # set "true" to apply the rules to the created or existing security group
+  #
+  #    # each rule in the rules list will be applied to every subject defined above
+  #    rules = [{
+  #      type = "ingress"
+  #      from_port = 433
+  #      to_port = 433
+  #      protocol = "tcp"
+  #      description = "Allow HTTPS ingress"
+  #    }]
+
+  type        = any
+  default     = []
+  description = <<-EOT
+    A convenient way to apply the same set of rules to a set of subjects. See README for details.
+    EOT
+}
+
+variable "security_group_create_timeout" {
+  type        = string
+  default     = "10m"
+  description = "How long to wait for the security group to be created."
+}
+
+variable "security_group_delete_timeout" {
+  type        = string
+  default     = "15m"
+  description = <<-EOT
+    How long to retry on `DependencyViolation` errors during security group deletion from
+    lingering ENIs left by certain AWS services such as Elastic Load Balancing.
+    EOT
+}
+
+variable "revoke_rules_on_delete" {
+  type        = bool
+  default     = false
+  description = <<-EOF
+    Instruct Terraform to revoke all of the Security Group's attached ingress and egress rules before deleting
+    the security group itself. This is normally not needed.
+    EOF
+}
+
+variable "vpc_id" {
+  type        = string
+  description = "The ID of the VPC where the Security Group will be created."
+}
+
+variable "inline_rules_enabled" {
+  type        = bool
+  default     = false
+  description = <<-EOF
+    NOT RECOMMENDED. Create rules "inline" instead of as separate `aws_security_group_rule` resources.
```

Comment:
> Perhaps there are use cases where it's absolutely needed

This is our thinking. If we do not provide the ability to use inline rules, someone will complain that they have a compelling need for it and ask us to provide that ability. They may even take the time and effort to implement it and provide a PR. We want to avoid that by both providing the ability and explaining why it is not a good idea. Also, if, down the road, the AWS provider improves on the current situation, this sets us up to take advantage of it without making a major change later.

As a side note, although `normalize.tf` is complex, it is also powerful and solves a lot of problems beyond the inline/discrete option.